spec/requests/bookmarks_controller_spec.rb

```diff
@@ -12,6 +12,58 @@
   end
 
   describe "#create" do
+    it "rate limits creates" do
+      SiteSetting.max_bookmarks_per_day = 1
+      RateLimiter.enable
+      RateLimiter.clear_all!
+
+      post "/bookmarks.json", params: {
+        post_id: bookmark_post.id,
+        reminder_type: "tomorrow",
+        reminder_at: (Time.zone.now + 1.day).iso8601
+      }
+
+      expect(response.status).to eq(200)
+
+      post "/bookmarks.json", params: {
+        post_id: Fabricate(:post).id
+      }
+      expect(response.status).to eq(429)
+      expect(response.parsed_body['errors']).to include(
+        I18n.t("rate_limiter.by_type.create_bookmark", time_left: "24 hours")
+      )
+    end
+
+    context "if the user reached the max bookmark limit" do
+      before do
+        @old_constant = Bookmark::BOOKMARK_LIMIT
+        Bookmark.send(:remove_const, "BOOKMARK_LIMIT")
+        Bookmark.const_set("BOOKMARK_LIMIT", 1)
+      end
+
+      it "returns failed JSON with a 400 error" do
+        post "/bookmarks.json", params: {
+          post_id: bookmark_post.id,
+          reminder_type: "tomorrow",
+          reminder_at: (Time.zone.now + 1.day).iso8601
+        }
+        post "/bookmarks.json", params: {
+          post_id: Fabricate(:post).id
+        }
+
+        expect(response.status).to eq(400)
+        user_bookmarks_url = "#{Discourse.base_url}/my/activity/bookmarks"
+        expect(response.parsed_body['errors']).to include(
+          I18n.t("bookmarks.errors.too_many", user_bookmarks_url: user_bookmarks_url)
+        )
+      end
+
+      after do
```

Comment:
I think it's the first time I see the `after` block at the end of a spec. We usually put it after the `before` block since it's "technically" part of the setup of the spec (more like the cleanup)